### DISCUSSION TITLE
Why is llm.eval_logits what it is?

### DISCUSSION THREAD

**Discussion by Koloth** [abetlen/llama-cpp-python#1850] [General]
I was recently trying to make a custom sampler ([documented on reddit here](https://www.reddit.com/r/LocalLLaMA/comments/1gy9qh9/making_custom_sampler_using_llamacpp_python/)) and I encountered what seems to be a seriously sharp edge in the high-level API that should at least be documented, if not changed.  Long story short: using `llm.eval_logits` to get the next token logits versus `llm.scores` leads to an approximately 15x slowdown (17 seconds to get a completion versus over 300 seconds) due to llm.eval_logits calling `to_list` (code [here](https://github.com/abetlen/llama-cpp-python/blob/7ecdd944624cbd49e4af0a5ce1aa402607d58dcc/llama_cpp/llama.py#L565)).  This was very surprising to me.  

Is there a reason that `eval_logits` is implemented in such an inefficient way, and essentially duplicates `scores`?  This took me far more debug time than I care to say to find!
